operation/operation.c

```diff
@@ -0,0 +1,271 @@
+            *time_unit = 'm';
+        }
+    }
+    return time_d;
+}
+
+// Log messages of different state to EEPROM, print to stdout and send to LORA
+void logf_msg(enum logs logEnum, oper_st *state, int n_args, ...)
+{
+    char msg[STRLEN_EEPROM];
+    char time_unit;
+    uint64_t time = get_time_with_decimal(&time_unit);
+    snprintf(msg, TIMESTAMP_LEN, "[%2llu,%01llu %c]",
+             time / DEC_OFFSET, time % DEC_OFFSET, time_unit);
+    char content[STRLEN_EEPROM - TIMESTAMP_LEN];
+    va_list args;
+    va_start(args, n_args);
+    vsnprintf(content, STRLEN_EEPROM - 1 - TIMESTAMP_LEN,
+              log_format[logEnum], args);
+    va_end(args);
+    strncat(msg, content, STRLEN_EEPROM - 1 - TIMESTAMP_LEN);
+
+    printf("%s\n", msg);
+    write_log_entry(msg, &state->eeprom_log_idx);
+
+    // values to store to eeprom
+    switch (logEnum)
+    {
+    case CALIB_COMPLETED:
+    case RECALIB_AFTER_REBOOT:
+        write_to_eeprom(COMP_INDEX_ADDR,
+                        &state->current_comp_idx, 1);
+        break;
+    case PILL_FOUND:
+        write_to_eeprom(PILLS_DETECTION_ADDR,
+                        &state->pills_detected, 1);
+    default:;
+    }
+
+    // LoRa messages to be sent to mqtt
+    switch (logEnum)
+    {
+    case BOOT:
+    case LORA_FAILED:
+        break;
+    default:
+        if (state->lora_connected)
+            send_msg_to_Lora(msg, false);
+    }
+}
+
+// generates operations state, data based on LoRa connection and EEPROM data
+oper_st init_operation()
+{
+    init_eeprom();
+    oper_st state;
+
+    state.eeprom_log_idx = get_stored_value(LOG_INDEX_ADDR);
+    logf_msg(BOOT, &state, 0);
+
+    state.current_comp_idx = get_stored_value(COMP_INDEX_ADDR);
+    state.pills_detected = get_stored_value(PILLS_DETECTION_ADDR);
+
+    if (state.current_comp_idx > PILLCOMP_COUNT)
+        state.current_comp_idx = PILLCOMP_COUNT;
+    if (state.pills_detected > PILLCOMP_COUNT)
+        state.pills_detected = 0;
+
+    init_Lora();
+    start_lora();
+    if ((state.lora_connected = connect_network()))
+    {
+        logf_msg(LORA_SUCCEED, &state, 0);
+    }
+    else
+    {
+        logf_msg(LORA_FAILED, &state, 0);
+    }
+    return state;
+}
+
+void print_state(oper_st state)
+{
+    printf("lora_connected: %s\n"
+           "eeprom_log_index: %hhu\n"
+           "current_comp_idx: %hhu\n"
+           "pills_detected: %hhu\n"
+           "\n",
+           state.lora_connected ? "true" : "false",
+           state.eeprom_log_idx,
+           state.current_comp_idx,
+           state.pills_detected);
+}
+
+// Loops infinitely until switch is pressed,
+// blinking LED while looping.
+// Leaves LED off after press.
+void blink_until_sw_pressed(SW *sw_proceed, LED *led, oper_st *state)
+{
+    logf_msg(WAITING_FOR_SW, state, 0);
+    uint8_t loop = 0;
+    while (!switch_pressed_debounced(sw_proceed))
+    {
+        if (loop++ == 0)
+            toggle_pwm(led);
+        loop %= 10;
```

Comment:
I think you got the gist of what we were thinking ; an addition, though:
Similarily as you did within the dispense loop, add the same read_log opportunity
1. here in the "blink loop"
2. in the wait_until_sw_pressed loop